provider/core/src/data_provider.rs

```diff
@@ -105,7 +104,74 @@ where
     T: ToOwned + ?Sized,
     <T as ToOwned>::Owned: Debug,
 {
-    pub cow: Cow<'d, T>,
+    pub(crate) cow: Cow<'d, T>,
+}
+
+impl<'d, T> DataPayload<'d, T>
+where
+    T: ToOwned + ?Sized,
+    <T as ToOwned>::Owned: Debug,
+{
+    /// Convert an owned Cow-compatible data struct into a DataPayload.
+    #[inline]
+    pub fn from_owned(data: <T as ToOwned>::Owned) -> Self {
+        Self {
+            cow: Cow::Owned(data),
+        }
+    }
+
+    /// Convert a borrowed Cow-compatible data struct into a DataPayload.
+    #[inline]
+    pub fn from_borrowed(data: &'d T) -> Self {
+        Self {
+            cow: Cow::Borrowed(data),
+        }
+    }
+
+    #[inline]
+    pub fn as_legacy_cow(&self) -> &Cow<'d, T> {
+        &self.cow
+    }
+
+    #[inline]
+    pub fn as_legacy_cow_mut(&mut self) -> &mut Cow<'d, T> {
```

Comment:
suggestion: next step is probably turning this into an `.as_mut(|&mut cow| {..})` closure-taking function, which we can support with Yoke

Reply:
Done.